File: ch05/src/cpp/simple_channel.cpp
#include <chrono>
#include <condition_variable>
#include <cstdint>
#include <deque>
#include <iostream>
#include <mutex>
#include <thread>
#include <vector>

template <typename T>
class Channel {
public:
    Channel() = default;

    void send(T message)
    {
        {
            std::lock_guard guard { mutex };
            queue.push_back(message);
        }
        cond.notify_one();
    }

    T receive()
    {
        std::unique_lock lock { mutex };
        cond.wait(lock, [this]() { return !queue.empty(); });
        const auto message = queue.front();
        queue.pop_front();
        return message;
    }

private:
    std::deque<T> queue;
    std::mutex mutex;
    std::condition_variable cond;
};

int main()
{
    auto c = Channel<std::int32_t> {};
    std::vector<std::jthread> handles;
    for (std::int32_t i { 0 }; i < 10; ++i) {
        handles.emplace_back([&c, i]() {
            const auto msg = c.receive();
            std::cout << "Thread " << i << " receives " << msg << std::endl;
        });
    }

    auto producer = std::jthread([&c]() {
        for (std::int32_t i { 0 }; i < 10; ++i) {
            c.send(i * i);
            std::this_thread::sleep_for(std::chrono::milliseconds { 100 });
        }
    });

    return 0;
}

File: ch05/src/simple_channel.rs
use std::collections::VecDeque;
use std::sync::{Condvar, Mutex};

pub struct Channel<T> {
    queue: Mutex<VecDeque<T>>,
    item_ready: Condvar,
}

impl<T> Channel<T> {
    pub fn new() -> Self {
        Self {
            queue: Mutex::new(VecDeque::new()),
            item_ready: Condvar::new(),
        }
    }

    pub fn send(&self, message: T) {
        // Push the message to the back of the queue
        // and notify one waiting receiver (if there are some)
        self.queue.lock().unwrap().push_back(message);
        self.item_ready.notify_one();
    }

    pub fn receive(&self) -> T {
        // Lock the mutex
        let mut b = self.queue.lock().unwrap();
        loop {
            // Pop one message from the queue, if there are some
            if let Some(message) = b.pop_front() {
                return message;
            }

            // Otherwise, block until notified
            // Note that the `wait` function unlock the mutex while waiting (and re-lock
            // before returning), so this does not block the mutex while waiting.
            b = self.item_ready.wait(b).unwrap();
        }
    }
}

impl<T> Default for Channel<T> {
    fn default() -> Self {
        Self::new()
    }
}
